### PR TITLE
DO NOT MERGE: Refactor NPM parser for V2 lock file

### PR DESCRIPTION
**Pull request by nishakm**

### CODE REVIEW

**Comment by cpanato** [comment]
there are some lints that should be addressed

File: npm/reader.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+// ParseManifestV2 converts a map[string]interface{} object into
+// a struct representing the NPM v2 lockfile
+func ParseManifestV2(data map[string]interface{}) (PackageLockV2, error) {
+	// PackageV2 and PackageLockV2 come from model.go
+	lock := PackageLockV2{}
+	// fill in Name, Version, LockfileVersion and Requires
+	lock.Name = data["name"].(string)
+	lock.Version = data["version"].(string)
+	lock.LockfileVersion = int(data["lockfileVersion"].(float64))
+	lock.Requires = data["requires"].(bool)
```
Comment (puerco) [inline npm/reader.go:58]:
```suggestion
	lock := PackageLockV2{
		Name: data["name"].(string),
		Version: data["version"].(string),
		LockfileVersion: int(data["lockfileVersion"].(float64)),
		Requires: data["requires"].(bool),
	}
```

File: npm/reader.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (c) 2022, Oracle and/or its affiliates.
```
Comment (puerco) [inline npm/reader.go:2]:
```suggestion
// Copyright 2023 The Linux Foundation and its contributors
```

File: npm/models.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (c) 2022, Oracle and/or its affiliates.
```
Comment (puerco) [inline npm/models.go:2]:
```suggestion
// Copyright 2023 The Linux Foundation and its contributors
```

**Review by puerco — COMMENTED** [review]
Thanks for looking into these @nishakm , some initial comments.

Can we rename the test directory to `testdata` to make the compiler ignore them? Also are we seprecating the old tests? If we're discarding them perhaps it makes more sense to simply delete them.

**Comment by puerco** [comment]
I just realized that the cargo package had some of these nits as well, so I fixed them in https://github.com/opensbom-generator/parsers/pull/36

File: npm/reader.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+// detectManifest detects what kind of manifest exists in the
+// provided path
+// and returns the path to the manifest that exists
+func DetectManifest(path string) string {
+	for i := range manifests {
+		fullPath := filepath.Join(path, manifests[i])
+		if _, err := os.Stat(fullPath); err == nil {
+			return fullPath
+		}
```
[the remark this answers is not in the document]
Comment (nishakm) [reply npm/reader.go:28]: I do return an empty string below.

**Comment by nishakm** [comment]
It looks like the tests are going to fail regardless. So I will just continue to update this branch until the lint passes and all the NPM tests pass.

**Comment by nishakm** [comment]
Redoing this since we've had a few changes merged

**Pull request closed without merging**